forecasttools/sbc.py

```diff
@@ -0,0 +1,206 @@
+            Which assumes a MCMC sampler e.g. NUTS.
+        seed : random.PRNGKey
+            Random seed.
+        kwargs : dict
+            Keyword arguments passed to `numpyro` models.
+        """
+        if sample_kwargs is None:
+            sample_kwargs = dict(
+                num_warmup=500, num_samples=100, progress_bar=False
+            )
+        if seed is None:
+            seed = random.PRNGKey(1234)
+        self.mcmc_kernel = mcmc_kernel
+        if not hasattr(mcmc_kernel, "model"):
+            raise ValueError(
+                "The `mcmc_kernel` must have a 'model' attribute."
+            )
+
+        self.model = mcmc_kernel.model
+        self.args = args
+        self.kwargs = kwargs
+        self.observed_vars = observed_vars
+
+        for key in self.observed_vars:
+            if key in self.kwargs and self.kwargs[key] is not None:
+                raise ValueError(
+                    f"The value for '{key}' in kwargs must be None for this to"
+                    " be a prior predictive check."
+                )
+
+        self.num_simulations = num_simulations
+        self.sample_kwargs = sample_kwargs
+
+        self.simulations = {}
+        self._simulations_complete = 0
+        prior_pred_rng, sampler_rng = random.split(seed)
+        self._prior_pred_rng = prior_pred_rng
+        self._sampler_rng = sampler_rng
+        self.num_samples = None
+
+    def _get_prior_predictive_samples(self):
+        """
+        Generate samples to use for the simulations by prior predictive
+        sampling. Then splits between observed and unobserved variables based
+        on the `observed_vars` attribute.
+        """
+        prior_predictive_fn = numpyro.infer.Predictive(
+            self.mcmc_kernel.model, num_samples=self.num_simulations
+        )
+        prior_predictions = prior_predictive_fn(
+            self._prior_pred_rng, *self.args, **self.kwargs
+        )
+        prior_pred = {
+            k: prior_predictions[v] for k, v in self.observed_vars.items()
+        }
+        prior = {
+            k: v
+            for k, v in prior_predictions.items()
+            if k not in self.observed_vars.values()
+        }
+        return prior, prior_pred
+
+    def _get_posterior_samples(self, seed, prior_predictive_draw):
+        """
+        Generate posterior samples conditioned to a prior predictive sample.
+        This returns the posterior samples and the number of samples. The
+        number of samples are used in scaling plotting and checking that each
+        inference draw has the same number of samples.
+        """
+        mcmc = MCMC(self.mcmc_kernel, **self.sample_kwargs)
+        obs_vars = {**self.kwargs, **prior_predictive_draw}
+        mcmc.run(seed, *self.args, **obs_vars)
+        num_samples = mcmc.num_samples
+        idata = az.from_numpyro(mcmc)
+        return idata, num_samples
+
+    def run_simulations(self):
```

Comment:
```suggestion
    def run_simulations(self) -> None:
```

(if choosing to follow theme of hinting)